oletools/util/log_helper/log_helper.py

```diff
@@ -0,0 +1,194 @@
+# 2018-02-07 v0.04 SA: - fixed control of handlers propagation
+# 2018-04-23 v0.05 SA: - refactored the whole logger to use an OOP approach
+
+# -----------------------------------------------------------------------------
+# TODO:
+
+
+from ._json_formatter import JsonFormatter
+from ._logger_adapter import OletoolsLoggerAdapter
+from . import _root_logger_wrapper
+import logging
+import sys
+
+
+LOG_LEVELS = {
+    'debug': logging.DEBUG,
+    'info': logging.INFO,
+    'warning': logging.WARNING,
+    'error': logging.ERROR,
+    'critical': logging.CRITICAL
+}
+
+DEFAULT_LOGGER_NAME = 'oletools'
+DEFAULT_MESSAGE_FORMAT = '%(levelname)-8s %(message)s'
+
+
+class LogHelper:
+    def __init__(self):
+        self._all_names = set()  # set so we do not have duplicates
+        self._use_json = False
+        self._is_enabled = False
+
+    def get_or_create_silent_logger(self, name=DEFAULT_LOGGER_NAME, level=logging.CRITICAL + 1):
+        """
+        Get a logger or create one if it doesn't exist, setting a NullHandler
+        as the handler (to avoid printing to the console).
+        By default we also use a higher logging level so every message will
+        be ignored.
+        This will prevent oletools from logging unnecessarily when being imported
+        from external tools.
+        """
+        return self._get_or_create_logger(name, level, logging.NullHandler())
+
+    def enable_logging(self, use_json, level, log_format=DEFAULT_MESSAGE_FORMAT, stream=None):
+        """
+        This function initializes the root logger and enables logging.
+        We set the level of the root logger to the one passed by calling logging.basicConfig.
+        We also set the level of every logger we created to 0 (logging.NOTSET), meaning that
+        the level of the root logger will be used to tell if messages should be logged.
+        Additionally, since our loggers use the NullHandler, they won't log anything themselves,
+        but due to having propagation enabled they will pass messages to the root logger,
+        which in turn will log to the stream set in this function.
+        Since the root logger is the one doing the work, when using JSON we set its formatter
+        so that every message logged is JSON-compatible.
+        """
+        if self._is_enabled:
+            raise ValueError('re-enabling logging. Not sure whether that is ok...')
+
+        log_level = LOG_LEVELS[level]
+        logging.basicConfig(level=log_level, format=log_format, stream=stream)
+        self._is_enabled = True
+
+        self._use_json = use_json
+        sys.excepthook = self._get_except_hook(sys.excepthook)
+
+        # since there could be loggers already created we go through all of them
+        # and set their levels to 0 so they will use the root logger's level
+        for name in self._all_names:
+            logger = self.get_or_create_silent_logger(name)
+            self._set_logger_level(logger, logging.NOTSET)
```

Comment:
When running as a script, we set the level of every logger we created to NOTSET, so they will use the root logger's level when deciding whether to log - keeping behavior consistent with what we [currently have](https://github.com/decalage2/oletools/blob/master/oletools/msodde.py#L999).